vita3k/codec/src/mjpeg.cpp

```diff
@@ -344,76 +365,136 @@ bool MjpegDecoderState::receive(uint8_t *data, DecoderSize *size) {
-
-            for (uint32_t y = 0; y < component_height; y++) {
-                uint8_t *dst = &data[channel_info[a].offset + y * component_width];
-                const uint8_t *src = frame->data[a] + (y < frame_component_height ? y * frame->linesize[a] : (frame_component_height - 1) * frame->linesize[a]);
-
-                if (y < frame_component_height) {
-                    std::memcpy(dst, src, frame_component_width);
-                    // Pad the rest with the last pixel value
-                    std::memset(dst + frame_component_width, dst[frame_component_width - 1], component_width - frame_component_width);
-                } else {
-                    // Copy the last row for padding
-                    std::memcpy(dst, dst - component_width, component_width);
+        // Implement YUV image align to pitch
+        for (int plane = 0; plane < 3; plane++) {
+            if (!frame->data[plane])
+                continue;
+
+            int src_width = frame->width;
+            int src_height = frame->height;
+
+            if (plane > 0) {
+                switch (color_space_out) {
+                case COLORSPACE_YUV444P:
+                    break;
+                case COLORSPACE_YUV440P:
+                    src_height /= 2;
+                    break;
+                case COLORSPACE_YUV441P:
+                    src_height /= 4;
+                    break;
+                case COLORSPACE_YUV422P:
+                    src_width /= 2;
+                    break;
+                case COLORSPACE_YUV420P:
+                    src_width /= 2;
+                    src_height /= 2;
+                    break;
+                case COLORSPACE_YUV411P:
+                    src_width /= 4;
+                    break;
+                case COLORSPACE_GRAYSCALE:
+                    break;
+                default:
+                    LOG_WARN("An attempt was made to use an unsupported color space.");
+                    return false;
+                }
+            }
+
+            for (int y = 0; y < src_height; y++) {
+                const uint8_t *src_row = frame->data[plane] + y * frame->linesize[plane];
+                uint8_t *dst_row = original_data + y * original_pitch[plane].x;
+
+                // Copy actual pixel data
+                std::memcpy(dst_row, src_row, src_width);
+
+                // Fill the rest with the last pixel value
+                uint8_t last_pixel = src_row[src_width - 1];
+                std::fill(dst_row + src_width, dst_row + original_pitch[plane].x, last_pixel);
+            }
+
+            // Fill remaining rows with the last row
+            const uint8_t *last_row = original_data + (src_height - 1) * original_pitch[plane].x;
+            for (int y = src_height; y < original_pitch[plane].y; y++) {
+                uint8_t *dst_row = original_data + y * original_pitch[plane].x;
+                std::memcpy(dst_row, last_row, original_pitch[plane].x);
+            }
```

Comment:
I think it is better to zero-fill instead (except if it is exactly the behavior of the PS Vita)

Reply:
That part is a loose emulation of the PS Vita's behavior.
In a typical JPEG decoding process, the JPEG is decoded in MCU units and then values outside the image's width/height range are discarded. However, the PS Vita doesn't discard values outside the pixel range in the decoded JPEG. Therefore, even if it's outside the image's width/height range, values are recorded in the decoding result if they're within the MCU block. Since it's difficult to exactly mimic this at a low level, I simulated the implementation through copying.
I question whether there are games where these values in this area have meaning (and there shouldn't be such games), but console games often make strange assumptions about hardware. So I guessed that the current approach is less likely to cause problems than zero-filling.